crates/ruma-common/src/events/receipt.rs

```diff
@@ -102,13 +104,149 @@ pub struct Receipt {
     /// The time when the receipt was sent.
     #[serde(skip_serializing_if = "Option::is_none")]
     pub ts: Option<MilliSecondsSinceUnixEpoch>,
+
+    /// The thread this receipt applies to.
+    #[serde(rename = "thread_id", skip_serializing_if = "crate::serde::is_default")]
+    pub thread: ReceiptThread,
 }
 
 impl Receipt {
     /// Creates a new `Receipt` with the given timestamp.
     ///
     /// To create an empty receipt instead, use [`Receipt::default`].
     pub fn new(ts: MilliSecondsSinceUnixEpoch) -> Self {
-        Self { ts: Some(ts) }
+        Self { ts: Some(ts), thread: ReceiptThread::Unthreaded }
+    }
+}
+
+/// The [thread a receipt applies to].
+///
+/// This type can hold an arbitrary string. To build this with a custom value, convert it from an
+/// `Option<string>` with `::from()` / `.into()`. [`ReceiptThread::Unthreaded`] can be constructed
+/// from `None`.
+///
+/// To check for values that are not available as a documented variant here, use its string
+/// representation, obtained through [`.as_str()`](Self::as_str()).
+///
+/// [thread a receipt applies to]: https://spec.matrix.org/v1.4/client-server-api/#threaded-read-receipts
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ReceiptThread {
+    /// The receipt applies to the timeline, regardless of threads.
+    ///
+    /// Used by clients that are not aware of threads.
+    ///
+    /// This is the default.
+    #[default]
+    Unthreaded,
+
+    /// The receipt applies to the main timeline.
+    ///
+    /// Used for events that don't belong to a thread.
+    Main,
+
+    /// The receipt applies to a thread.
+    ///
+    /// Used for events that belong to a thread with the given thread root.
+    Thread(OwnedEventId),
+
+    #[doc(hidden)]
+    _Custom(PrivOwnedStr),
+}
+
+impl ReceiptThread {
+    /// Get the string representation of this `ReceiptThread`.
+    ///
+    /// [`ReceiptThread::Unthreaded`] returns `None`.
+    pub fn as_str(&self) -> Option<&str> {
+        match self {
+            Self::Unthreaded => None,
+            Self::Main => Some("main"),
+            Self::Thread(event_id) => Some(event_id.as_str()),
+            Self::_Custom(s) => Some(&s.0),
+        }
+    }
+}
+
+impl<T> From<Option<T>> for ReceiptThread
+where
+    T: AsRef<str> + Into<Box<str>>,
+{
+    fn from(s: Option<T>) -> Self {
+        match s {
+            None => Self::Unthreaded,
+            Some(s) => match s.as_ref() {
+                "main" => Self::Main,
+                s_ref => match EventId::parse(s_ref) {
+                    Ok(event_id) => Self::Thread(event_id),
+                    Err(_) => Self::_Custom(PrivOwnedStr(s.into())),
+                },
```

Comment:
This seems a little weird for me, though I see the idea. Maybe the match on whether the parse succeeded could be a match on the first char instead, going to `Self::Thread` if the first char is `$`? Would that be in line with the spec?

Reply:
The spec only says that it's supposed to be an event ID for the moment, so your idea probably makes sense too.

In your suggestion, after the `match`, if the `EventId` parsing fails it's an error, right?

Reply:
Yeah.